server/services/questionsToDelete.ts

```diff
@@ -0,0 +1,115 @@
+   * Determines which of the currently answered questions need to be deleted
+   *
+   * This is done by traversing the questionnaire tree (from the start) and
+   * returning the question IDs which were not visited, given the current
+   * answers.
+   *
+   * @param report config
+   * @param answeredQuestions list of questions answered (and their responses)
+   *
+   * @returns list of question IDs to delete
+   */
+  public static forGivenAnswers(config: IncidentTypeConfiguration, answeredQuestions: AnsweredQuestion[]): string[] {
+    const currentQuestionsIds = answeredQuestions.map(question => question.code)
+
+    const visitedQuestions = this.traverseQuestionnaire(config, config.startingQuestionId, answeredQuestions)
+
+    // Delete questions not visited, they're in non-visited branches
+    return currentQuestionsIds.filter(questionId => !visitedQuestions.includes(questionId))
+  }
+
+  /**
+   * Traverses the questionnaire tree and return the visited question IDs
+   *
+   * The traversal starts at `start` and is based off the already answered
+   * questions.
+   *
+   * @param report config
+   * @param start question ID where to start
+   * @param answeredQuestions list of questions answered (and their responses)
+   *
+   * @returns list of question IDs to delete
+   */
+  private static traverseQuestionnaire(
+    config: IncidentTypeConfiguration,
+    start: string,
+    answeredQuestions: AnsweredQuestion[],
+  ): string[] {
+    const nextQuestionId = this.findNextQuestionId(config, start, answeredQuestions)
+    if (!nextQuestionId) {
+      return []
+    }
+
+    return [start, ...this.traverseQuestionnaire(config, nextQuestionId, answeredQuestions)]
+  }
+
+  /**
+   * Returns the next question ID given the current answers
+   *
+   * The logic is "clever" enough to determine the next question ID
+   * for some unanswered questions if no branching occurs.
+   *
+   * The implications of this is that more of the previous answers can
+   * be potentially preserved. For example, changing an answer may leads to
+   * a new branch with some unanswered questions, however this branch
+   * may not have any further sub-branching and it may re-join a previously
+   * answered question 'QX': In this case this question 'QX' would still be
+   * valid and it could be retained.
+   *
+   * @param report config
+   * @param start question ID where to start
+   * @param answeredQuestions list of questions answered (and their responses)
+   *
+   * @returns the next question ID after a question
+   */
+  private static findNextQuestionId(
+    config: IncidentTypeConfiguration,
+    start: string,
+    answeredQuestions: AnsweredQuestion[],
+  ): string | null {
+    const answeredQuestion = answeredQuestions.find(question => question.code === start)
+    let questionConfig = null
+    let answerConfig = null
+
+    if (answeredQuestion) {
+      questionConfig = config.questions[answeredQuestion.code]
+      if (questionConfig) {
```

Comment:
if `questionConfig` is **not** found for an answered question, this answer essentially becomes unvisited, right? (and therefore would get deleted)

Reply:
This is a bit of an edge case but I covered it in case for whatever reason the report had some questions which we cannot find in the config.

It would then try the else branch, not find the question config there neither and eventually leave the if/else and return null as start can't be visited (so yes, it would ultimately delete the questions with id `start`)

Reply:
there is no else branch to no `questionConfig`

Reply:
Yes it does work. Sorry you are right, I didn't look at the code properly when answering you, so yeah it does not enter the `else` branch of course.

However notice how critically I'm declaring the `questionConfig` and `answerConfig` variables before the the branch and they're `null` unless we can correctly find the config/answer and unambiguously determine the next question to visit.

So at the end of the day, if we can't find the question config or the question was not answered and there is no branching (or if it's the last question!) this method would return null and end end the traversal. The end result is, as you said, that this unknown question would not be visited and it will be deleted.

Hope this makes sense and it's less confusing this time 😅

Reply:
great, that's what i thought. i like that self-healing aspect of this cos it's always someone feasible for a rando-question to appear in the answer set… a weird sync error; some data race; technically-erroneous-according-to-config nomis data…